test/library/packages/csv/CSV.chpl

```diff
@@ -0,0 +1,167 @@
+module CSV {
+  record CSVIO {
+    var ch: channel;
+    var sep: string = ",";
+    var hasHeader = true;
+
+    class ReaderWriterMismatchError: Error {}
+
+    proc init(filename: string, reader:bool=true, sep=",") {
+      if reader then ch = openreader(filename);
+                else ch = openwriter(filename);
+    }
+    proc init(ch: channel, sep: string=",", hasHeader=true) {
+      this.ch = ch;
+      this.sep = sep;
+      this.hasHeader = hasHeader;
+    }
+
+    iter readCSV(type t) throws where isRecordType(t) {
+      use Reflection;
+      var r: t;
+      var skipHeader = hasHeader;
+
+      if ch.writing {
+        throw new owned ReaderWriterMismatchError();
+      }
+
+      for l in ch.lines() {
+        const line = l.strip(leading=false);
+        if line.length == 0 then
+          continue;
+        const vals = line.split(sep);
+        for param i in 1..numFields(t) {
+          getFieldRef(r, i) = vals[i]: getField(r, i).type;
+        }
+        if skipHeader {
+          skipHeader = false;
+        } else {
+          yield r;
+        }
+      }
+    }
+
+    iter readCSV(type t...) throws {
```

Comment:
It seems a little odd for the `readCSV` tuple overload to take var-args, requiring tuple expansion from the user, whereas the `writeRecordCSV` just takes a tuple. Of the two, I think requiring a tuple would be better for the interface, especially if we want to add additional arguments to these methods in the future.

Reply:
I think the common case for `readCSV` would be to explicitly list the types expected in the `readCSV` call, not to put them into a tuple first.  If you mean tuple expansion is required for the `24*real` trick, it still would be needed if readCSV took a tuple and you wanted to add any types beyond the 24 reals.

```
// expand the 24 reals, put them into a tuple with the string.
readCSV(((...(24*real)), string));
```

`writeCSV` writes the things that `readCSV` returns.

Reply:
Right. I was thinking of the case where you define your column types outside of the read/write call, which I'd expect to be fairly common:

```
type t = (real, real, real, string);

// Current interface requires this:
readCSV(...t);

// Accepting tuples would look like this:
readCSV(t);
```

It could be valuable to have others weigh in on this as well.